Add source location to task and tasktrace object

radoslawcybulski opened the pull request:
Adds source location of next resume point to `task` objects. When constructing `tasktrace` object, source location of resume points will be added to it. When printing, the source location of next resume point will be printed as well, hopefully improving debug experience.

Example:
```
0x2b3899 /home/y/work/scylladb-test/build/Debug/seastar/libseastar.so+0x454c364 ...
   --------
   test_app/main.cpp:15:15
   seastar::continuation<seastar::internal::promise_base_with_type<void>, ...
```

Partially fixes https://github.com/scylladb/seastar/issues/2381

Commit:
Add a std::source_location (resume_point) to task object.

Add an empty, default constructed std::source_location object
to the task object and getter / setter.

Commit:
Add calls to update resume_point

Add calls to update `resume_point` variable with location of
next resume location to all `await_suspend` functions and `then`
functions.

Commit:
Update backtrace with source locations of resume points

Add resume point locations to `tasktrace` object.
Update `formatter::format` to print source location of next resume alone
with task type.

avikivity commented on include/seastar/core/task.hh:
```diff
@@ -35,6 +36,8 @@ class task {
     scheduling_group _sg;
 private:
+    std::source_location resume_point = {};
+
```
This is making a very heavily used object larger.

radoslawcybulski replied:
After adding `std::source_location`

`coroutine_traits_base<>::promise_type` (which inherits from task) is 64 bytes,
`continuation<Promise, Func, Wrapper, T>` is around 104-120, depending on exact version.

No difference except for that single unlucky value, who is "promoted" to the next 64 byte cache line in case of `continuation`.

avikivity commented:
Aren't those addresses derivable from the stack trace? Or are those points lost?

How large is std::source_location?

radoslawcybulski commented:
> Aren't those addresses derivable from the stack trace? Or are those points lost?
Probably with a little bit of magic. You'd still have to store a pointer somewhere, because stacktrace is valid for an actually running task, with my patch adds "next instruction" to tasks, which are suspended.
 
> How large is std::source_location?
8 bytes.

radoslawcybulski commented:
my patch
```
instructions_per_op: mean=48465.87 standard-deviation=33.03 median=48459.42 median-absolute-deviation=22.63 maximum=48520.81 minimum=48437.81
instructions_per_op: mean=48482.90 standard-deviation=21.16 median=48487.16 median-absolute-deviation=6.97 maximum=48504.13 minimum=48447.34
instructions_per_op: mean=48469.42 standard-deviation=18.34 median=48472.45 median-absolute-deviation=17.26 maximum=48488.53 minimum=48448.81
```

master
```
instructions_per_op: mean=48337.41 standard-deviation=21.29 median=48334.08 median-absolute-deviation=19.67 maximum=48364.48 minimum=48317.02
instructions_per_op: mean=48352.36 standard-deviation=27.87 median=48347.35 median-absolute-deviation=16.28 maximum=48395.26 minimum=48322.29
instructions_per_op: mean=48339.83 standard-deviation=26.48 median=48323.12 median-absolute-deviation=20.80 maximum=48372.64 minimum=48319.03
```

My patch will run more instructions, as there are additional stores to the task object (std::source_location needs to be written).

Why do we measure instructions per operation? We are not billed by instruction count, rather by time, we should measure total running time instead? Does it work as a sanity check or am i missing something?

avikivity commented:
> 
> Why do we measure instructions per operation? We are not billed by instruction count, rather by time, we should measure total running time instead? Does it work as a sanity check or am i missing something?


Time is very unstable (just try it) since it depends on temperature and cooling.

Instructions is a more stable proxy for time (but inaccurate since instructions-per-clock can change).

avikivity commented:
So we add 0.25% overhead. Hard to judge if it's worthwhile.

avikivity commented:
> > Time is very unstable (just try it) since it depends on temperature and cooling.
> 
> Agree.
> 
> > Instructions is a more stable proxy for time (but inaccurate since instructions-per-clock can change).
> 
> It's absolutely stable, but we're counting steps, which - in my opinion - sort-of blinds us. It's like two different walks, the same step count, but step sizes differ, so is distance. Here for example if we're running at max IPC, then we've 0.25% overhead. If otherwise we're waiting for memory, then the write is pretty much free (we've already waited for this cache line, because other stuff is read / written there) and proc is waiting for other memory pieces, so it can do our write.

Writes are more-or-less free, except for the increase in instruction footprint (which are extra reads).

We aren't close to running at max IPC. Typical for mini-benchmarks like perf-simple-query is 2, and for production ~1. The main killer is waiting for instruction fetch (although maybe in production mispredicts would also contribute since the the data is less regular). I'm sure an M4 would give much higher IPC.

tchaikov commented on include/seastar/core/coroutine.hh:
```diff
@@ -30,6 +30,7 @@
 
 #ifndef SEASTAR_MODULE
 #include <coroutine>
+#include <source_location>
```
it took me a while to check if it's safe to use `<source_location>` instead of `util/std-compat.hh` . the answer is yes. as we support the latest two major versions of the GCC and Clang compilers. in general, we use libstdc++, which added the support to `source_location` in https://github.com/gcc-mirror/gcc/commit/57d76ee9cf6265e012fad6286adfaeaba9414c11, which is included by gcc 12 and up. and the latest stable release of GCC is GCC 14.2.

nyh replied:
Opened https://github.com/scylladb/seastar/issues/2961 on this.

tchaikov reviewed (commented):
if the performance impact could be a concern, can we make it an optional feature which is enabled only in the debug builds?

radoslawcybulski commented:
> if the performance impact could be a concern, can we make it an optional feature which is enabled only in the debug builds?

This has been discussed here (https://github.com/scylladb/seastar/issues/2381) and consensus is that it's the release is where it's useful the most.

tgrabiec commented:
> > > Aren't those addresses derivable from the stack trace? Or are those points lost?
> > 
> > 
> > Probably with a little bit of magic. You'd still have to store a pointer somewhere, because stacktrace is valid for an actually running task, while my patch adds "next instruction" to tasks, which are suspended.
> > > How large is std::source_location?
> > 
> > 
> > 8 bytes.
> 
> Maybe we can tolerate it. @tgrabiec what's your opinion?

I think it's probably ok.

tchaikov reviewed (approved):
lgtm

bitpathfinder commented on include/seastar/core/task.hh:
```diff
@@ -35,6 +36,8 @@ class task {
 protected:
     scheduling_group _sg;
 private:
+    std::source_location resume_point = {};
```
I'd put the `resume_point` first since it's larger than `scheduling_group`. that way the `task` object size is 4 bytes less.

radoslawcybulski replied:
Done.

tgrabiec replied:
Not really - if the class is not packed then the size includes the alignment, so order doesn't matter.

bitpathfinder replied:
Yes, that's true. The object will be aligned to 8 bytes boundary. However in case we add something after these variables, less than 5 bytes, then there will be a difference, so having it ordered by size might be beneficial in the future.

radoslawcybulski commented on include/seastar/core/task.hh:
```diff
@@ -26,14 +26,18 @@
 
 SEASTAR_MODULE_EXPORT
 class task {
+    std::source_location resume_point = {};
```
Done.

radoslawcybulski commented:
@scylladb/seastar-maint please merge.

denesb commented:
> So we add 0.25% overhead. Hard to judge if it's worthwhile.

@avikivity I think the gain is much higher than 0.25% here. We need to reduce the bar of entry into debugging seastar applications.

denesb commented:
> @denesb can you tell me, how did you figure out performance differences for my patch (or any patch)? Sub 1% precision looks amazing to me, how did you do that?

I didn't, I just reacted to Avi's comment. I didn't claim that the perf regression is something else, just that 0.25% that Avi claimed is worth it for the gain of seeing line numbers.

michoecho commented:
> @denesb can you tell me, how did you figure out performance differences for my patch (or any patch)? Sub 1% precision looks amazing to me, how did you do that?

We're talking about the `perf_simple_query` results posted earlier in this thread.

radoslawcybulski commented:
@scylladb/seastar-maint  shameless ping - please merge

radoslawcybulski commented:
@scylladb/seastar-maint shameless ping - please merge.

nyh commented:
> @scylladb/seastar-maint shameless ping - please merge.

@avikivity and @tgrabiec are both maintainers and were involved in this review, I think they should get the honor of deciding to merge it.

Since this is the first time I'm seeing this issue, I have a beginner's question - where are all those fairly big "source location" strings saved in the executable (I'm not talking about the 8-byte pointers, which are also a problem - I'm talking about the actual strings). Can you estimate how much the stripped Scylla executable will grow due to these strings? I don't have any ideas if there are hundreds such strings, or hundreds of thousands.

avikivity commented:
There's a gratuitous merge commit at the end.

@travisdowns please opine on whether this would be terrible for performance for you, or great for debugging.

avikivity commented:
> > There's a gratuitous merge commit at the end.
> > @travisdowns please opine on whether this would be terrible for performance for you, or great for debugging.
> 
> @avikivity If we don't want to pessimize existing applications, we can always hide the stuff added in this patch behind the preprocessor.

That's the easy way out, but I don't want to take it.

> But please don't just let this patch rot. Let me have it at least in dev mode for starters. I've wasted so much time over the years on tests where the main debugging difficulty was finding out where a continuation came from.

That's reasonable. Let's see if we can push it to release too.

nyh commented:
> Weren't the tradition dictated by a need to be more obscure due to a partially closed source nature of past scyllas?

No, I wasn't talking about Scylla's tradition, but in general - addr2line isn't specific to Scylla or closed source.
In any case, addr2line does not need the source code - just an executable with extra (or separate) "debugging information" (https://en.wikipedia.org/wiki/DWARF) which helps convert instruction-pointer (a.k.a. PC) addresses to file name, line number and function name. As I said, I agree that storing all this stuff in the executable and allowing the executable to introspect it is certainly more convenient than addr2line, but in the old days it wasn't done because it has a cost, as you noticed. But we can definitely decide that the convenience is worth the small cost.

avikivity commented:
The binary size growth isn't very interesting. This is a system that often manages dozens of terabytes per node. Nobody will notice a megabyte of strings, there are seven orders of magnitude separating the two.

However, it does generate an idea - can we use __builtin_return_address() here? It gives a tiny bit more information, and won't grow the executable.

radoslawcybulski commented:
Yes, i think so.

Note, that source code information is `compile-time` stuff (it will survive optimizations / inlining), while `__builtin_return_address()` is runtime (it will tell us, where return address is after optimizations, which is roughly what we have now).

avikivity commented:
> I stand corrected, stack traces are actually sort of useful now, debug info has improved massively.
> 
> > Stack traces (when explored with addr2line -i) show much more information. Each return address on the stack encodes all the inlined call-sites that led to it up to the next uninlined caller. The more inlining happens, the more information it carries.
> 
> This assumes every instruction has unambiguous inlined sequence of calls leading to it. Or am i missing something?


It doesn't assume anything.

If inlining happens to duplicate a call to (let's say) future::then, then you will get more information.

If it doesn't, then you don't.

In the worst case __builtin_return_address will return the same amount of information as std::source_location.

In the common case, it will return more, since you will see the direct caller, and its callers up to the first uninlined function.

> 
> > See https://godbolt.org/z/G678Eo7rh. Although there is just one location that calls g(), it knows exactly which path called it through a.
> 
> Good example. I took the liberty of removing `[[gnu::noinline]]` and `[[gnu::always_inline]]` and adding `h` implementation, that prints the `ptr` value. The code repeats 4 different values at `-O2` and `-O3` (which makes sense as clang just inlined to some depth). Adding flags used for `release` build makes the code print single value. I also added a small `if` test, that calls `foo` from both branches - prints the same value in both cases.
> 
> Am i missing something? It doesn't take much to generate code, that drops ambiguous return pointers after optimization. Is there a way to resolve those i don't know (which would make me a X)? Otherwise that's pretty much the point i'm so poorly making. You can probably make it sort-of work with clever `noinline` flags (adding `noinline` to functions that also call `__builtin_return_address` would probably do it). Why bother? This will also hurt performance.

If you compare to std::source_location, then in all combinations of inlining options it returns one call site. __builtin_return_address will return from 1 to 32 call sites, depending on the options.

Here's the example updated to report both std::source_location and __builtin_return_address. If you change h() to print both, you will get 32 different values for __builtin_return_address but only one value for std::source_location.

https://godbolt.org/z/MqaocP1x5

michoecho commented:
> This might be improvable (although i doubt i'm the first to think about it), if - instead of calling __builtin_return_address (which gets return address of current function, which is not that useful or we need to stop inlining) we could get next execution address pointer. This sort-of reverves the issue - now we have to force all then implementations. Then next instruction pointer will be as if return pointer of then call if it were forced to noinline, we should get missing top level inlined frames and it should fly.
My instinct tells me i'm not the first one to think about it and they are major issues with this approach. While std::source_location will work 100% of times on all systems.

@radoslawcybulski You mean, use `rip` instead of `std::source_location` like I did in https://github.com/scylladb/seastar/issues/2381#issuecomment-2263125627?

I wanted to use the instruction pointer instead of source location from the start, because it theoretically gives more information, but I didn't try to push that because in practice (when I was privately using that patch for debugging) the compiler was very often generating bad debug info for the addresses I obtained this way. (E.g. pointing to the right file, but to line 0 instead of the actual line). `std::source_location` prevents problems like that. So it's a tradeoff — less info, but more reliable, or more info with a chance that the info will be less useful (without manual investigation to find the actual line) because the compiler didn't care enough to generate good debug info.

radoslawcybulski commented:
> @radoslawcybulski You mean, use `rip` instead of `std::source_location` like I did in [#2381 (comment)](https://github.com/scylladb/seastar/issues/2381#issuecomment-2263125627)?

Yep, the same. Also your patch highlights my other point - it's rare to figure out something new, most of the time if people are not using something there's a reason for it.

radoslawcybulski commented on include/seastar/core/slim_source_location.hh:
```diff
@@ -0,0 +1,73 @@
+
+        [[gnu::always_inline]] slim_source_location(const char* file = __builtin_FILE(), std::int32_t line = __builtin_LINE(), std::int32_t column = __builtin_COLUMN())
+                : _file(file), _line(line), _column(column) {
+#ifdef SSL_HAS_RIP_X86
```
slim source line.
My brain keeps telling me i've seen this abbreviation, but i've not found anything better. I can just expand into `SLIM_SOURCE_LINE_***`, those macros are contained in this file anyway.

avikivity commented on include/seastar/core/slim_source_location.hh:
```diff
@@ -0,0 +1,73 @@
+                : _file(file), _line(line), _column(column) {
+#ifdef SSL_HAS_RIP_X86
+            std::uintptr_t rip;
+            asm("leaq 0(%%rip), %0":"=r"(rip));
```
We could slim it down with something like

```c++
struct slim_data* sd;
asm ("1: lea 2f, %0 \n"
          ".pushsection .rodata.something \n"
          "2: \n"
          ".quad 1b \n"
          ".quad %1 \n"
          ".long %2, %3 \n"
          ".popsection"
          : "=r"(sd) : "i"(__builtin_FILE()) : "i"(__builtin_LINE()), "i"(__builtin_COLUMN()));
```

This pushes the data to the data section and we end up with one instruction.

radoslawcybulski replied:
Well, i do understand few identifiers from the example, but that's all. :)

What is `1:` and `2:` (i assume labels? They don't seem used? Are those global identifiers or local to function / module?). Why we do `.quad 1b` (`quad` is 64bit probably, so `1b` is just `1`)? `long` is 32bit i assume?  What is `=r...` syntax?

avikivity replied:
It's black magic and if you ever manage to read it fluently, you're cursed for life.

avikivity replied:
Right. I don't think it will work.

I guess we can fall back to the standard source_location.

radoslawcybulski replied:
@michoecho why this woulnd't work?

```
struct slim_data* sd;

void *foo(const char *file, int line, int column) {
asm ("1: lea 2f, %0 \n"
          ".pushsection .rodata.something \n"
          "2: \n"
          ".quad 1b \n"
          ".quad %1 \n"
          ".long %2, %3 \n"
          ".popsection"
          : "=r"(sd) : "i"(file) : "i"(line), "i"(column));
...
}
```
?

michoecho replied:
> Why this woulnd't work?

Because Avi's wish would be to combine the instruction pointer and the source location (and maybe `__PRETTY_FUNCTION__` too, why not) into a static, constant-initialized struct, so that all that statically-known info can be passed around inside tasks as a single pointer at runtime. And that's what the piece of asm is trying to accomplish.

But if you want that struct to be constant-initialized, you must initialize it with constants. And in your example, you are trying to initialize it with not-constants. (If the function is always inlined, then the values of `line` and `column` and `file` are known at compile time, but something like this doesn't fit into the type system. A function must stand on its own, there's no "inline-only constexpr"). It will be rejected by the compiler. (The `"i"` in `"i"(line)` stands for "immediate", that is: a constant woven into the instruction. And you are trying to pass a variable there).

> why do we need this black magic incantation with ` __attribute__((section("..."), used))`?

I want all `tracepoint_entry` structs to be collected by the linker into their own section (`tracepoints` here) so that my trace decoder can later, while preparing for decoding time, look at the Scylla binary and iterate over all `tracepoint_entry` structs scattered inside it, to get the signatures for all event types in the trace.

If they aren't packed into a section, they will be in random places in `.rodata`, and they can't be found without knowing their names. If they are packed into a section, my decode can open the Scylla executable, find its `tracepoints` section, cast it to a `tracepoint_entry[]` array, iterate over it, and generate a corresponding decoder for each `TRACEPOINT` in the program. 

For example, I can put 
```c++
TRACEPOINT(event_level::debug, "io_begin", "class", io_request.class, "id", io_request.id, "size", io_request.size);
```
anywhere in the program, and it will produce a
```
tracepoint_entry{.name = "io_begin", .signature = "class:u8,id:u64,size:u64", ...}
```
to the `tracepoints` section at some address X, and the decoder will iterate over X (among others), see the signature and use it generate a decoder function which reads a
```
struct io_begin {
    uint8_t class;
    uint64_t id;
    uint64_t size;
};
```
from the trace whenever the header is X.
And then the functions generated from signatures can be compiled into the actual decoder program.

Other sections are not needed for anything, they are just there to keep the metadata strings neatly ordered in the ELF.

The `used` is probably not needed either, I just wanted to make sure the linker won't play garbage collection tricks on me.

---

Anyway, all that's not very applicable in this thread, because macros are off the table.

michoecho commented:
> > compiler was very often generating bad debug info for the addresses I obtained this way. (E.g. pointing to the right file, but to line 0 instead of the actual line).
> 
> Was this with addr2line or llvm-addr2line? We recently saw that llvm-addr2line is better.

@avikivity I'm 80% sure that the problem was at compile time, not decode time. But I don't remember my problem well, I only have the vague memory that I was getting the RIPs I wanted but the file and line info for those RIPs in the DWARF was wrong(/useless) sometimes.

gusev-p reviewed (commented):
@radoslawcybulski @scylladb/scylla-maint I see that all maintainers are ok with this approach, lets push this forward, this promises to be very helpful for debugging.

I didn't know about this work, so I [invented my own](https://github.com/scylladb/seastar/pull/3255). I wanted to improve reactor reporting in `do_dump_task_queue` because `typeid(task)->name()` is useless in most cases, especially for coroutines. I wanted to avoid inflating the task by another 8 bytes, so I tried to extract a meaningful symbol from existing data structures instead. We can use resume address stored in the coroutine frame, although it won't tell us the exact co_await within the coroutine. This would probably have been enough to diagnose the reactor queue problems, but if we're ok to store the source_location in the task instance it would be much better.

> std::source_location works, but is not always useful. It can point to, say, some function in `loop.hh` and give you not much idea about what's going on.

We can gradually improve this in the future by capturing `source_location` in the public functions of `loop.hh` and forwarding them to the appropriate task instance.

@radoslawcybulski please update the `do_dump_task_queue` to `use task->get_resume_point()`.

Commit:
Use source_location in `do_dump_task_queue`

Add printing source_location of resume point for each task,
that `do_dump_task_queue` logs.

radoslawcybulski commented on include/seastar/util/backtrace.hh:
```diff
@@ -159,14 +159,17 @@ class tasktrace {
 public:
     using entry = std::variant<shared_backtrace, task_entry>;
     using vector_type = boost::container::static_vector<entry, 16>;
+    using vector_resume_points_type = boost::container::static_vector<std::source_location, 16>;
```
Updated.

radoslawcybulski commented on include/seastar/util/backtrace.hh:
```diff
@@ -159,14 +159,17 @@ class tasktrace {
 public:
     tasktrace() = default;
     tasktrace(simple_backtrace main, vector_type prev, size_t prev_hash, scheduling_group sg) noexcept;
+    tasktrace(simple_backtrace main, vector_type prev, vector_resume_points_type prev_resume_points, size_t prev_hash, scheduling_group sg) noexcept;
```
Done.

radoslawcybulski commented:
```
/__w/seastar/seastar/apps/rpc_tester/rpc_tester.cc:574:42: internal compiler error: in make_decl_rtl, at varasm.cc:1459
```

yay...

EDIT: it's this one https://gcc.gnu.org/bugzilla/show_bug.cgi?format=multiple&id=121643 . We need a different version of gcc or clang.

gusev-p commented:
> ```
> /__w/seastar/seastar/apps/rpc_tester/rpc_tester.cc:574:42: internal compiler error: in make_decl_rtl, at varasm.cc:1459
> ```
> 
> yay...
> 
> EDIT: it's this one https://gcc.gnu.org/bugzilla/show_bug.cgi?format=multiple&id=121643 . We need a different version of gcc or clang.

Maybe the compiler would like the `await_transform` more?

https://stackoverflow.com/questions/76110225/why-is-promise-typeawait-transform-greedy
> If you want to have certain specific co_awaits be instrumented by the promise, you can do that by doing co_await my_instrumented_awaitable(<expr>) manually as needed. But if you need all of them to be instrumented (which can be used for tracing suspend/resume points), if it's not something you pick-and-choose, then you use await_transform.

gusev-p commented on src/util/backtrace.cc:
nit: can be made written more compactly:
```
for (const auto & [ e, resume_loc ] = b._prev) {
}
```

gusev-p commented on include/seastar/core/coroutine.hh:
```diff
@@ -218,7 +220,8 @@ public:
     template<typename U>
-    void await_suspend(std::coroutine_handle<U> hndl) noexcept {
+    void await_suspend(std::coroutine_handle<U> hndl, std::source_location sl = std::source_location::current()) noexcept {
+        hndl.promise().update_resume_point(sl);
```
I mean if we have a lot of `co_await`-s for already resolved futures we will have many redundant `MOV` instructions -- these resume points won't be needed.

gusev-p commented:
> what is more important to know - who called `sleep_abortable` or where within `sleep_abortable` the future will be resumed? We probably don't want to keep two source locations for both cases.

Probably we can assume that the target audience of the `source_location` feature are programmers who use Seastar, not Seastar developers -> we should capture source_location at the Seastar api boundaries.

gusev-p committed:
reactor: improve do_dump_task_queue reporting

gusev-p commented:
@radoslawcybulski I've [tuned reactor/do_dump_task_queue](https://github.com/gusev-p/seastar/commit/9721bfcaeda6ab5268f6385492ddb5d8bf29ed14) a bit, please merge into your PR.

I like the result:
```
WARN  2026-02-13 23:08:51,953 [shard 0:main] seastar - Too long queue accumulated for main (21 tasks)
 1: tasks/task_manager.cc:253:9
        tasks/task_manager.cc:220:8
 1: tasks/task_manager.cc:220:8
 6: seastar/src/core/future-util.cc:133:16
        seastar/include/seastar/core/future.hh:1653:54
        tasks/task_manager.cc:327:12
 7: N7seastar13shared_futureIJEE12shared_stateE
 6: replica/distributed_loader.cc:419:5
        replica/distributed_loader.cc:331:9
        replica/distributed_loader.cc:453:13
        N7seastar9coroutine17parallel_for_eachIZN7replica18distributed_loader17populate_keyspaceERNS_7shardedINS2_8databaseEEERNS4_IN2db15system_keyspaceEEERNS2_8keyspaceENS_13basic_sstringIcjLj15ELb1EEEE3$_0EE
        replica/distributed_loader.cc:441:14
        N7seastar12_GLOBAL__N_116thread_wake_taskE
        seastar/include/seastar/core/thread.hh:261:28
        seastar/include/seastar/core/thread.hh:263:12
        N7seastar12_GLOBAL__N_116thread_wake_taskE
        seastar/include/seastar/core/thread.hh:261:28
        seastar/include/seastar/core/thread.hh:263:12
        seastar/src/core/app-template.cc:165:37
        seastar/src/core/app-template.cc:167:12
        seastar/src/core/app-template.cc:169:12
```

Let's review/merge this!

gusev-p reviewed (commented):
LGTM

gusev-p committed:
coroutine.hh: disable source_location for GCC to avoid ICE

radoslawcybulski commented:
@avikivity @scylladb/seastar-maint please merge.

gusev-p commented:
@avikivity ping, please review

gusev-p commented on include/seastar/core/coroutine.hh:
```diff
@@ -194,8 +193,8 @@ public:
-    void await_suspend(std::coroutine_handle<U> hndl, std::source_location sl = std::source_location::current()) noexcept {
-        hndl.promise().update_resume_point(sl);
+    void await_suspend(std::coroutine_handle<U> hndl SEASTAR_COROUTINE_LOC_PARAM) noexcept {
+        SEASTAR_COROUTINE_LOC_STORE(hndl.promise());
```
> Please link the gcc bug in the changelog entry.

It's linked in the definitions of these macros. Do you want it to be mentioned in the commit message?

> Does this have to be a macro?

We need to conditionally add an optional parameter and pass it to `update_resume_point`. I'm not sure how to do that without the macro. One option is to declare two versions of `await_suspend` -- one with the source_location and one without it. But then we'll need to do that for all `await_suspend` declarations, there are quite a few of them.

nvartolomei commented on include/seastar/util/std-compat.hh:
```diff
@@ -56,3 +56,18 @@ using source_location
+
+#if defined(__GNUC__) && !defined(__clang__)
+// GCC Workaround: Strip source_location to prevent ICE during RTL expansion.
+// See: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=114675
```
Wrong link?

> 114675 - warning for "reference to not fully constructed object" - P3 enhancement
